src/ecc_key.c

```diff
@@ -0,0 +1,1528 @@
+
+	rc = 0;
+ret:
+	rv = pthread_mutex_unlock(&ec_key->lock);
+	assert(rv == 0);
+	DEBUG("return %d (%s)", rc, zpc_error_string(rc));
+	return rc;
+}
+
+int zpc_ec_key_import(struct zpc_ec_key *ec_key, const unsigned char *buf,
+				unsigned int buflen)
+{
+	target_t target;
+	int rc, rv, seclen;
+	size_t i, trailing_spki_len = 0;
+
+	UNUSED(rv);
+
+	if (pkeyfd < 0) {
+		return ZPC_ERROR_DEVPKEY;
+	}
+	if (ec_key == NULL) {
+		return ZPC_ERROR_ARG1NULL;
+	}
+	if (buf == NULL) {
+		return ZPC_ERROR_ARG2NULL;
+	}
+
+	if (buflen < MIN_EC_BLOB_SIZE || buflen > MAX_EC_BLOB_SIZE) {
+		return ZPC_ERROR_ARG3RANGE;
+	}
+
+	rv = pthread_mutex_lock(&ec_key->lock);
+	assert(rv == 0);
+
+	if (ec_key->refcount != 1) {
+		rc = ZPC_ERROR_OBJINUSE;
+		goto ret;
+	}
+
+	if (ec_key->curve_set != 1) {
+		rc = ZPC_ERROR_EC_CURVE_NOTSET;
+		goto ret;
+	}
+	if (ec_key->type_set != 1) {
+		rc = ZPC_ERROR_KEYTYPENOTSET;
+		goto ret;
+	}
+
+	if (ec_key->type == ZPC_EC_KEY_TYPE_CCA && !is_cca_ec_key(buf, buflen)) {
+		rc = ZPC_ERROR_EC_NO_CCA_SECUREKEY_TOKEN;
+		goto ret;
+	}
+
+	if (ec_key->type == ZPC_EC_KEY_TYPE_EP11 && !is_ep11_ec_key_with_header(buf, buflen)) {
+		rc = ZPC_ERROR_EC_NO_EP11_SECUREKEY_TOKEN;
+		goto ret;
+	}
+
+	if (!ec_key_blob_has_valid_mkvp(ec_key, buf)) {
+		rc = ZPC_ERROR_WKVPMISMATCH;
+		goto ret;
+	}
+
+	if (!ec_key_blob_is_pkey_extractable(ec_key, buf)) {
+		rc = ZPC_ERROR_BLOB_NOT_PKEY_EXTRACTABLE;
+		goto ret;
+	}
+
+	/* In case of ep11, the imported buffer may contain the actual secure key
+	 * blob concatenated with a public key spki. */
+	if (ec_key->type == ZPC_EC_KEY_TYPE_EP11)
+		trailing_spki_len = buflen - ep11_get_raw_blob_length(buf);
+
+	/* Set (secure) private key. Host lib not needed for this. */
+	seclen = buflen - trailing_spki_len;
+	memset(ec_key->cur.sec, 0, sizeof(ec_key->cur.sec));
+	memcpy(ec_key->cur.sec, buf, seclen);
+	ec_key->cur.seclen = seclen;
+	ec_key->key_set = 1;
+
+	/* Extract and set public key. For this we need the related host lib. If
+	 * the host lib is not available, only the secure key is available in
+	 * this key object. */
+	if (ec_key->type == ZPC_EC_KEY_TYPE_CCA) {
+		if (!swcaps.ecdsa_cca) {
+			rc = 0;
+			goto ret;
+		}
+		ec_key->pubkey_set = 1;
+		rv = pthread_mutex_lock(&ccalock);
+		assert(rv == 0);
+		rc = ec_key_extract_public_cca(&cca,
+					(unsigned char *)&ec_key->cur.sec, ec_key->cur.seclen,
+					(unsigned char *)&ec_key->pub.pubkey, &ec_key->pub.publen,
+					true);
+		rv = pthread_mutex_unlock(&ccalock);
+		assert(rv == 0);
+		if (rc != 0 || ec_key->pub.publen == 0)
+			ec_key->pubkey_set = 0;
+	} else {
+		if (!swcaps.ecdsa_ep11) {
+			rc = 0;
+			goto ret;
+		}
+		ec_key->pubkey_set = 1;
+		rv = pthread_mutex_lock(&ep11lock);
+		assert(rv == 0);
+		for (i = 0; i < ec_key->napqns; i++) {
+			rc = get_ep11_target_for_apqn(&ep11, ec_key->apqns[i].card,
+					ec_key->apqns[i].domain, &target, true);
+			if (rc)
+				continue;
+			rc = ec_key_extract_public_ep11(&ep11, ec_key->curve,
```

Comment:
Hm, so in ec_key_extract_public_ep11 you obtain the MACed SPKI, but you only use the EC-point at this time. 
First question: Why do you obtain the MACed SPKI then, you only really need the raw SPKI for that.
Second question: Don't you also want to have the MACed SPKI for such a key and keep in in the key object? So instead of extracting just the EC point, pass back the whole MACed-SPKI?

Reply:
Good point! Code changed.